ow_faults_detection/src/FaultDetector.cpp

```diff
@@ -0,0 +1,141 @@
+                                                  10,
+                                                  &FaultDetector::powerTemperatureListener,
+                                                  this);
+}
+
+// Creating Fault Messages
+template<typename fault_msg>
+void FaultDetector::setFaultsMessageHeader(fault_msg& msg){
+  msg.header.stamp = ros::Time::now();
+  msg.header.frame_id = "world";
+}
+
+template<typename bitsetFaultsMsg, typename bitmask>
+void FaultDetector::setBitsetFaultsMessage(bitsetFaultsMsg& msg, bitmask bm) {
+  setFaultsMessageHeader(msg);
+  msg.value = bm.to_ullong();
+}
+
+template<typename fault_msg>
+void FaultDetector::setComponentFaultsMessage(fault_msg& msg, ComponentFaults value) {
+  setFaultsMessageHeader(msg);
+  msg.value = static_cast<uint>(value);
+}
+
+// publish system messages
+void FaultDetector::publishSystemFaultsMessage(){
+  ow_faults_injection::SystemFaults system_faults_msg;
+  setBitsetFaultsMessage(system_faults_msg, m_system_faults_bitset);
+  m_system_fault_msg_pub.publish(system_faults_msg);
+}
+
+//// Publish Camera Messages
+void FaultDetector::cameraTriggerPublishCb(const ros::TimerEvent& t){
+  ow_faults_injection::CamFaults camera_faults_msg;
+  auto diff = m_cam_raw_time - m_cam_trigger_time;
+  if (m_cam_trigger_time <= m_cam_raw_time  
+    &&  m_cam_raw_time <= m_cam_trigger_time + ros::Duration(2) 
+    || diff < ros::Duration(0) && ros::Duration(-1) < diff) {
+    m_system_faults_bitset &= ~isCamExecutionError;
+  } else {
+    m_system_faults_bitset |= isCamExecutionError;
+    setComponentFaultsMessage(camera_faults_msg, ComponentFaults::Hardware);
+  }
+
+  publishSystemFaultsMessage();
+  m_camera_fault_msg_pub.publish(camera_faults_msg);
+}
+
+//// Publish Power Faults Messages
+void FaultDetector::publishPowerSystemFault(){
+  ow_faults_injection::PowerFaults power_faults_msg;
+  //update if fault
+  if (m_temperature_fault || m_soc_fault) {
+    //system
+    m_system_faults_bitset |= isPowerSystemFault;
+    //power
+    setComponentFaultsMessage(power_faults_msg, ComponentFaults::Hardware);
+  } else {
+    m_system_faults_bitset &= ~isPowerSystemFault;
+  }
+  publishSystemFaultsMessage();
+  m_power_fault_msg_pub.publish(power_faults_msg);
+}
+
+// Listeners
+//// Camera listeners
+void FaultDetector::camerTriggerCb(const std_msgs::Empty& msg){
+  m_cam_trigger_time = ros::Time::now();
+}
+
+void FaultDetector::cameraRawCb(const sensor_msgs::Image& msg){
+  m_cam_raw_time = ros::Time::now();
+}
+
+//// Power Topic Listeners
+void FaultDetector::powerTemperatureListener(const std_msgs::Float64& msg)
+{
+  m_temperature_fault = msg.data > THERMAL_MAX;
+  publishPowerSystemFault();
+}
+
+void FaultDetector::powerSOCListener(const std_msgs::Float64& msg)
+{
+  float newSOC = msg.data;
+  if (isnan(m_last_SOC)){
+    m_last_SOC = newSOC;
+  }
+  m_soc_fault = ((newSOC <= SOC_MIN)  ||
```

Comment:
So the fault is not detected until the SOC is at minimum?  By this time the battery is effectively dead.  I thought that detection at this level was to occur immediately after injection.

Reply:
see above comment